cli/start: log uid/gid details in the log file

knz opened the pull request:
Fixes #21342.
(I used a non-obtrusive, non-prescriptive solution since root processes are used in containers.)

In troubleshooting scenarios where users report "I am encountering
permission errors" or "I can't seem to access files", etc., we will
want to check that they are running with consistent uid/gid
information. So log them upon start up.

In addition, an informative disclaimer is also logged if the process
happens to be running as root, for example:

```
W180109 17:28:31.252473 1 cli/start.go:773  process is starting as root.
DISCLAIMER: CockroachDB running as root can pose a security risk to the rest of the system unless securely contained.
```

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/21346)
<!-- Reviewable:end -->

knz committed:
cli/start: log uid/gid details in the log file

In troubleshooting scenarios where users report "I am encountering
permission errors" or "I can't seem to access files", etc., we will
want to check that they are running with consistent uid/gid
information. So log them upon start up.

Release note: None

a-robinson commented on pkg/cli/start.go:
```diff
@@ -761,6 +761,19 @@ func reportConfiguration(ctx context.Context) {
+	if syscall.Geteuid() == 0 {
+		// If a human ever looks at the logs, tell them what we think.
+		log.Warning(ctx, "process is starting as root.\n"+
+			"DISCLAIMER: CockroachDB running as root can pose a security risk to the rest of the system unless securely contained.")
```
This message creates worry without any explanation -- it claims that what the user is doing is insecure without saying why or helping the user make an informed decision about what to do. They'll naturally ask "what's insecure about it?" and may get more afraid than is necessary, thinking it opens their system open to intrusion by outsiders or something.

knz replied:
Given we haven't audited CockrochDB thoroughly, it seems fair to disclaim any expectation of security.

a-robinson replied:
Does that mean we should always log a security warning regardless of what user the process is being run as? After all, the user the process is being run as has access to all the data in the database.

knz replied:
I don't care much about the fact the user has access to the data in the database :)

What I care about (and what the proposed warning highlights) is that a root process has potential access to every other process and user **in the entire system**, not just this database.

bdarnell replied:
I agree with Alex. On a system that runs a database, the contents of that database are likely the most important data on that system. In past jobs if someone were to gain access to the `mysql` user on one of my database machines, I wouldn't take any comfort in the fact that it wasn't `root`. Similarly, if I run cockroach as user `bdarnell`, it would have access to a lot of non-database stuff that probably concerns me more than anything that would require root. 

In some cases, there are particular reasons why you might want to be picky about things like this. For example, there have been security vulnerabilities if mysql has write access to its own config file, so you may not want to run mysql as root *even in a container*. 

Our security philosophy has been to take precautions against vulnerabilities like this without relying solely on the process's permissions. For example, we will not write to disk paths specified over the network if they are outside the directory specified by `--external-io-dir`, even if the process has permissions to do so. Running as non-root is a best practice and an extra layer of protection, but it's no more necessary for cockroach than it is for anything else. 

If we do think that running as root is a risk, logging a warning in an already very noisy log is not a very good solution. We could have the server simply refuse to start as root without an override flag. Both postgres and mysql apparently do this. I don't think this is a good idea given the prevalence these days of running as root in containers, but if we want to take a stand against running as root this would be the way to do it.

knz replied:
I think I'd prefer to prevent from running as root except if there's an override.
But that would be a more invasive change (needing changes to docs etc) which I am not ready to steward at this time. 

So I removed the warning for now.

PTAL

knz commented:
TFYR!
